src/casp/services/api/services/consensus_engine/strategies.py

```diff
@@ -0,0 +1,230 @@
+            raise ValueError(
+                "`cell_ontology_term_id_to_cell_type` mapping is not found in the cell ontology resource dictionary"
+            )
+
+        self.ancestors_dictionary = cell_ontology_resource_dict["ancestors_dictionary"]
+        self.ontology_term_id_to_name_dict = cell_ontology_resource_dict["cell_ontology_term_id_to_cell_type"]
+
+
+class OntologyAwareConsensusStrategy(ConsensusStrategyProtocol):
+    """
+    Handle ontology-aware consensus strategy, summarizing query neighbor context by cell type ontology. Weights are
+    assigned to each neighbor cell type based on their distance and cell type ontology, and the weights are propagated
+    to their ancestors in the cell type ontology graph.
+    Algorithm:
+
+    1. Get Metadata for each unique neighbor cell.
+    2. Iterate over each query cell.
+       2.1. Get distances for each neighbor cell.
+       2.2. Calculate weights for each neighbor cell.
+       2.3. Update weights for each neighbor cell and their ancestors in the cell type ontology.
+       2.4. Normalize the weights if required.
+
+    :param normalize: Boolean flag indicating whether to normalize the weights.
+    :param prune_threshold: Threshold for pruning weights below a certain value. If 0, no pruning is performed.
+    :param cell_ontology_resource: Cell ontology resource object.
+    :param cell_operations_dm: Cell operations data manager object.
+    """
+
+    REQUIRED_CELL_INFO_FEATURE_NAMES = ["cas_cell_index", "cell_type", "cell_type_ontology_term_id"]
+
+    def __init__(
+        self,
+        normalize: bool = True,
+        prune_threshold: float = 0.1,
+        cell_ontology_resource: t.Optional[CellOntologyResource] = None,
+        cell_operations_dm: t.Optional[CellOperationsDataManager] = None,
+    ):
+        self.cell_ontology_resource = cell_ontology_resource or CellOntologyResource()
+        self.cell_operations_dm = cell_operations_dm or CellOperationsDataManager()
+        self.normalize = normalize
+        self.prune_threshold = prune_threshold
+
+    def _get_ontology_aware_neighborhood_context(
+        self,
+        neighbors: t.List[MatchResult.Neighbor],
+        neighbors_metadata_dict: t.Dict[str, schemas.CellariumCellMetadata],
+    ) -> t.List[schemas.NeighborhoodSummaryOntologyAware]:
+        """
+        Utilize the ontology-aware method to assign weights to neighbor cells based on their distance and cell type
+        ontology, to inform context summarization.
+
+        :param neighbors: Neighbors of the query cell as determined by the matching engine.
+        :param neighbors_metadata_dict: Metadata for each neighbor cell.
+        :return: A list of `schemas.AnnotationInfoOntologyAware` instances, each representing the weighted cell type
+             ontology term for a neighbor cell.
+        """
+
+        neighbor_distances = np.asarray([neighbor.distance for neighbor in neighbors])
```

Comment:
We need to be careful with distances and similarity scores. It looks like Vertex AI is returning cosine similarity scores. The distance, in that case, is typically defined as `1 - cossim(x, y)`. If we are storing _distance_ as an attribute of a match, we need to make sure that it is indeed a distance. My workaround has been to change the sign of `gamma`, which is a terrible idea. The better thing to do is to fix this upstream. Can you check what is going on? Is Vertex AI actually returning similarity scores and calling it distance, or have we called it distance by mistake? Does Vertex AI always return similarity scores regardless of what distance metric is used (likely), or returns distance for some metrics and similarity scores for other metrics (unlikely)?

Reply:
MatchNeighbor object from google's sdk has a variable distance, however in some tutorial notebooks they mention that distance is a similarity. I am going to send them an email and see what they respond. I will leave this discussion unresolved before we make sure it is either distance or a similarity.